letsencrypt-compatibility-test/letsencrypt_compatibility_test/test_driver.py

```diff
@@ -0,0 +1,324 @@
+
+    domains = list(plugin.get_testable_domain_names())
+    success = test_deploy_cert(plugin, temp_dir, domains)
+
+    if success and args.enhance:
+        success = test_enhancements(plugin, domains)
+
+    good_rollback = test_rollback(plugin, config, backup)
+    return names_match and success and good_rollback
+
+
+def test_deploy_cert(plugin, temp_dir, domains):
+    """Tests deploy_cert returning True if the tests are successful"""
+    cert = crypto_util.gen_ss_cert(util.KEY, domains)
+    cert_path = os.path.join(temp_dir, "cert.pem")
+    with open(cert_path, "w") as f:
+        f.write(OpenSSL.crypto.dump_certificate(
+            OpenSSL.crypto.FILETYPE_PEM, cert))
+
+    for domain in domains:
+        try:
+            plugin.deploy_cert(domain, cert_path, util.KEY_PATH)
+        except le_errors.Error as error:
+            logger.error("Plugin failed to deploy ceritificate for %s:", domain)
+            logger.exception(error)
+            return False
+
+    if not _save_and_restart(plugin, "deployed"):
+        return False
+
+    verify_cert = validator.Validator().certificate
+    success = True
+    for domain in domains:
+        if not verify_cert(cert, domain, "127.0.0.1", plugin.https_port):
+            logger.error("Could not verify certificate for domain %s", domain)
+            success = False
+
+    if success:
+        logger.info("HTTPS validation succeeded")
+
+    return success
+
+
+def test_enhancements(plugin, domains):
+    """Tests supported enhancements returning True if successful"""
+    supported = plugin.supported_enhancements()
+
+    if "redirect" not in supported:
+        return True
+
+    for domain in domains:
+        try:
+            plugin.enhance(domain, "redirect")
+        except le_errors.Error as error:
+            logger.error("Plugin failed to enable redirect for %s:", domain)
+            logger.exception(error)
+            return False
+
+    if not _save_and_restart(plugin, "enhanced"):
+        return False
+
+    verify_redirect = functools.partial(
+        validator.Validator().redirect, "localhost", plugin.http_port)
+    success = True
+    for domain in domains:
+        if not verify_redirect(headers={"Host" : domain}):
+            logger.error("Improper redirect for domain %s", domain)
+            success = False
+
+    if success:
+        logger.info("Enhancments test succeeded")
+
+    return success
+
+
+def _save_and_restart(plugin, title=None):
+    """Saves and restart the plugin, returning True if no errors occurred"""
+    try:
+        plugin.save(title)
+        plugin.restart()
+        return True
+    except le_errors.Error as error:
+        logger.error("Plugin failed to save and restart server:")
+        logger.exception(error)
+        return False
+
+
+def test_rollback(plugin, config, backup):
+    """Tests the rollback checkpoints function"""
+    try:
+        plugin.rollback_checkpoints(2)
```

Comment:
Any particular reason 2 was chosen? My plan (for Apache) is to have a full checkpoint at the end of ```deploy_certificate``` and for each enhancement.  If we have more than redirect than this might break.

If the goal is simply to revert everything a large number can be used and rollback_checkpoints shouldn't have a problem with it.